File: 2025-1/sesion-de-clase-20/SoftProg/SoftProgNegocio/src/main/java/pe/edu/pucp/inf30/softprog/boimpl/OrdenVentaBOImpl.java
package pe.edu.pucp.inf30.softprog.boimpl;

import java.io.IOException;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;
import pe.edu.pucp.inf30.softprog.bo.Estado;
import pe.edu.pucp.inf30.softprog.bo.IOrdenVentaBO;
import pe.edu.pucp.inf30.softprog.bo.transaccion.TransaccionalBO;
import pe.edu.pucp.inf30.softprog.bo.transaccion.comando.ComandoGuardarOrdenVenta;
import pe.edu.pucp.inf30.softprog.config.DBManager;
import pe.edu.pucp.inf30.softprog.dao.logistica.ventas.ILineaOrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.dao.logistica.ventas.IOrdenVentaDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.logistica.ventas.LineaOrdenVentaDAOImpl;
import pe.edu.pucp.inf30.softprog.daoimpl.logistica.ventas.OrdenVentaDAOImpl;
import pe.edu.pucp.inf30.softprog.model.logistica.ventas.OrdenVenta;

/**
 *
 * @author eric
 */
public class OrdenVentaBOImpl 
        extends TransaccionalBO 
        implements IOrdenVentaBO 
{
    private final IOrdenVentaDAO ordenVentaDao;
    private final ILineaOrdenVentaDAO lineaOrdenVentaDao;
    
    public OrdenVentaBOImpl() {
        this.ordenVentaDao = new OrdenVentaDAOImpl();
        this.lineaOrdenVentaDao = new LineaOrdenVentaDAOImpl();
    }
    
    @Override
    public List<OrdenVenta> listar() {
        return this.ordenVentaDao.listar();
    }

    @Override
    public OrdenVenta obtener(int id) {
        return this.ordenVentaDao.buscar(id);
    }

    @Override
    public void guardar(OrdenVenta modelo, Estado estado) {
        ComandoGuardarOrdenVenta comando = 
                new ComandoGuardarOrdenVenta(
            this.ordenVentaDao, 
            this.lineaOrdenVentaDao, 
            estado, 
            modelo
        );
        this.ejecutarTransaccion(comando);
    }
    
    @Override
    public void eliminar(int id) {
        throw new UnsupportedOperationException("Not supported yet.");
    }
}

File: 2025-1/sesion-de-clase-20/SoftProgNet/SoftProgNegocio/BOImpl/OrdenVentaBOImpl.cs
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using PUCP.Edu.Pe.SoftProg.Config;
using PUCP.Edu.Pe.SoftProg.Modelo.Logistica.Ventas;
using PUCP.Edu.Pe.SoftProg.Negocio.BO;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAO.Logistica.Ventas;
using PUCP.Edu.Pe.SoftProg.Persistencia.DAOImpl.Logistica.Ventas;

namespace PUCP.Edu.Pe.SoftProg.Negocio.BOImpl {
    public class OrdenVentaBOImpl : IOrdenVentaBO {
        private readonly IOrdenVentaDAO ordenVentaDAO;
        private readonly ILineaOrdenVentaDAO lineaOrdenVentaDAO;

        public OrdenVentaBOImpl() {
            this.ordenVentaDAO = new OrdenVentaDAOImpl();
            this.lineaOrdenVentaDAO = new LineaOrdenVentaDAOImpl();
        }

        public void Guardar(OrdenVenta ordenVenta, Estado estado) {
            if (estado == Estado.Nuevo) {
                DbTransaction transaction = null;
                using (DbConnection conexion = DBManager.GetInstance().GetConnection()) {
                    try {
                        transaction = conexion.BeginTransaction();

                        int idOrden = this.ordenVentaDAO.Insertar(ordenVenta, conexion, transaction);

                        foreach (LineaOrdenVenta linea in ordenVenta.LineasOrdenVenta) {
                            linea.OrdenVenta = new OrdenVenta {
                                Id = idOrden,
                            };
                            this.lineaOrdenVentaDAO.Insertar(linea, conexion, transaction);
                        }

                        transaction.Commit();
                    }
                    catch (Exception) {
                        transaction.Rollback();
                    }
                }
            }
            else {
                DbTransaction transaction = null;
                using (DbConnection conexion = DBManager.GetInstance().GetConnection()) {
                    try {
                        transaction = conexion.BeginTransaction();

                        this.ordenVentaDAO.Modificar(ordenVenta);

                        foreach (LineaOrdenVenta linea in ordenVenta.LineasOrdenVenta) {
                            this.lineaOrdenVentaDAO.Modificar(linea);
                        }

                        transaction.Commit();
                    }
                    catch (Exception) {
                        transaction.Rollback();
                    }
                }
            }
        }

        public void Eliminar(int id) {
            DbTransaction transaction = null;
            using (DbConnection conexion = DBManager.GetInstance().GetConnection()) {
                try {
                    transaction = conexion.BeginTransaction();

                    OrdenVenta ordenVenta = this.ordenVentaDAO.Buscar(id);

                    foreach (LineaOrdenVenta linea in ordenVenta.LineasOrdenVenta) {
                        this.lineaOrdenVentaDAO.Eliminar(linea.Id);
                    }

                    this.ordenVentaDAO.Eliminar(ordenVenta.Id);

                    transaction.Commit();
                }
                catch (Exception e) {
                    transaction.Rollback();
                }
            }
        }

        public OrdenVenta Obtener(int id) {
            return this.ordenVentaDAO.Buscar(id);
        }

        public List<OrdenVenta> Listar() {
            return this.ordenVentaDAO.Listar();
        }
    }
}
